### PR TITLE
Add bumper sensor (code, message, GUI, etc.)

### PR DESCRIPTION
**Pull request by longjie**
This PR is for adding bumper sensor feature.
It's my first time to modify STDR code, so please check if it is reasonable.
The added code is based on the sonar sensor.

### CODE REVIEW

**Comment by etsardou** [comment]
Hi @longjie and thanks for the contribution! We will have a look at it as soon as possible ;)

**Comment by etsardou** [comment]
Hey @longjie, I have just tried the code and everything seems to run as should. Great work! The only thing missing is the bumper support in the robot creator GUI window (press the small button with the yellow dot in GUI). The files needed to be changed are the ones located here:

> stdr_simulator/stdr_gui/src/stdr_gui/stdr_robot_creator/
> stdr_simulator/stdr_gui/include/stdr_gui/stdr_robot_creator/

and the writer files in the parser (`stdr_parser_xml_writer` and `stdr_parser_yaml_writer` cpp & h).

This way a user will be able to add a bumper to a new robot from the GUI as well (and not only by altering the yaml or xml files).

Right now I am a little tight on time. Can you handle these changes?

Again, thank you for the contribution!

File: stdr_robot/CMakeLists.txt [hunk 1]
```diff
@@ -66,6 +66,10 @@ add_library(stdr_laser src/sensors/laser.cpp)
 add_dependencies(stdr_laser stdr_msgs_gencpp) # wait for stdr_msgs to be build
 target_link_libraries(stdr_laser ${catkin_LIBRARIES} stdr_sensor_base)
 
+add_library(stdr_bumper src/sensors/bumper.cpp)
```
Comment (czalidis) [inline stdr_robot/CMakeLists.txt:69]: This library needs to be added to the install targets as well.

File: stdr_gui/launch/stdr_gui.launch [hunk 2]
```diff
@@ -1,5 +1,5 @@
 <launch>
 
-	<node name="$(anon stdr_gui_node)" pkg="stdr_gui" type="stdr_gui_node"/>
+	<node name="$(anon stdr_gui_node)" pkg="stdr_gui" type="stdr_gui_node" output="screen"/>
```
Comment (czalidis) [inline stdr_gui/launch/stdr_gui.launch]: `output="screen"` is probably unnecessary.

**Comment by longjie** [comment]
Hi, thanks,
I see the robot creator's change. Many changes seem to be needed. I may take rather long time to fill them all. Would you merge the bumper feature prior to the GUI support? Thank you.

**Comment by corot** [comment]
Hi, my :+1: to merge this PR before updating the GUI. It works fine, and the waiting will make harder the merge (it is already tangled with the kinematic model)

However, the bumper specifications need some clarification... to me radius and pose seem like contradictory parameters. I finally moved to radius, because my first attempt placing the bumper's pose in its physical position and radius to a small value didn't work (the bumper arc becomes very small). Now I just set x and y as 0 and place bumpers with "polar" coordinates, just theta and radius. But I must specify a radius to make the bumpers well outside my robot footprint to make them work . Here is my config for a TurtleBot2:

```
    # Bumpers
    - bumper:
        bumper_specifications:
          frequency: 20.0
          frame_id: bumpers/front
          contact_angle: 1.15
          radius: 0.22
          pose:
            x: 0.0
            y: 0.0
#            x: 0.175
#            y: 0.0
            theta: 0.0
    - bumper:
        bumper_specifications:
          frequency: 20.0
          frame_id: bumpers/left
          contact_angle: 1.0
          radius: 0.22
          pose:
            x: 0.0
            y: 0.0
#            x: 0.0838994692557
#            y: 0.15357694833
            theta: 1.07080
    - bumper:
        bumper_specifications:
          frequency: 20.0
          frame_id: bumpers/right
          contact_angle: 1.0
          radius: 0.22
          pose:
            x: 0.0
            y: 0.0
#            x: 0.0838994692557
#            y: -0.15357694833
            theta: -1.07080
```

Note that I specify a radius of 0.22, while the robot radius is 0.175.

Another hint is: should not it use sensor_msgs/Ranger with fixed distance instead of creating a Bumper message? Main advantage is that move_base costmaps accept sensor_msgs/Ranger messages to mark obstacles, so no further processing is needed to use bumpers on navi stack. In fact sensor_msgs/Range is the defacto standard message type for range sensors (IR, ultrasonic) including fixed-range sensors such as bumpers.

**Comment by etsardou** [comment]
Hey all! I agree to merge this PR without the GUI part (the robot creation one), since its functional and we clearly have no time to properly implement it..

Concerning the Range comment I agree, since this will provide support for the costmaps package (even though many of the Range parameters will be unused). 

So I propose to merge this and open another issue to implement bumpers with Range messages. @czalidis ?

**Comment by czalidis** [comment]
I think we should merge this feature when it is implemented correctly using Range messages. I see no point at merging this PR and immediately after dropping support for the Bumper.msg, that it introduces, and switch to Range messages. 

In any case this PR has conflicts and cannot be merged directly, so I guess the person who will resolve the conflicts (or just rebase) can make the change to Range messages as well. I would do it myself, but I have no time to spare for the next month.

@longjie as the original author of this PR, can you resolve the conflicts and make the change to `sensor_msgs/Range`, in order for this PR to get merged?

Keep in mind that according to [range_sensor_layer](https://github.com/DLu/navigation_layers) for costmap, the convention for fixed distance rangers is that `min_range == max_range` and range values can be `-Inf (== object detected)` or `Inf (== no object detected)`.

**Comment by longjie** [comment]
OK, I'll check if I can do as you mentioned.
Closing this PR for now.
Thanks.

**Comment by etsardou** [comment]
Reopened just to keep track of it ;)

**Comment by czalidis** [comment]
Closing this PR, since development for this feature has been moved on branch [feature/bumper-sensor](https://github.com/stdr-simulator-ros-pkg/stdr_simulator/tree/feature/bumper-sensor).

**Pull request closed without merging**